### PR TITLE
[INTEL MKL] Using MKL DNN sgemm

### PR DESCRIPTION
**Pull request by agramesh1**
Added code to call MKL DNN sgemm when building from MKL DNN opensources only.  This PR depends on https://github.com/tensorflow/tensorflow/pull/21011, so it should be merged only after 21011 is merged

### CODE REVIEW

File: tensorflow/core/kernels/matmul_op.cc [hunk 1]
```diff
@@ -591,6 +591,19 @@ TF_CALL_bfloat16(REGISTER_CPU);
 TF_CALL_int32(REGISTER_CPU);
 TF_CALL_complex64(REGISTER_CPU_EIGEN);
 TF_CALL_complex128(REGISTER_CPU_EIGEN);
+
+#elif defined(INTEL_MKL) && defined(DO_NOT_USE_ML)
+
+// MKL opensource supports only float for matrix-multiplication
+TF_CALL_float(REGISTER_CPU_EIGEN);
```
Comment (rmlarsen) [inline tensorflow/core/kernels/matmul_op.cc]: Why do we register with the "eigen" label to call MKL instead of Eigen?  Confusing...

**Comment by agramesh1** [comment]
@rmlarsen It looks like there are some unit tests that specifically require Eigen. I think a change was made by someone at Google to add the Eigen registrations. https://github.com/tensorflow/tensorflow/commit/d8697935d334bb0f2e1c9bccfe9a2a7bee9785cc . For MKL accelerated matmuls we define Eigen version as well default cpu which uses MKL.

**Comment by rmlarsen** [comment]
@agramesh1 Sounds good. I'll try to remember why that is on my own time ;-)

File: tensorflow/core/kernels/matmul_op.cc [hunk 2]
```diff
@@ -591,6 +591,19 @@ TF_CALL_bfloat16(REGISTER_CPU);
...
+TF_CALL_double(REGISTER_CPU);
+TF_CALL_half(REGISTER_CPU);
+TF_CALL_bfloat16(REGISTER_CPU);
+
+TF_CALL_int32(REGISTER_CPU);
```
Comment (tatianashp) [inline tensorflow/core/kernels/matmul_op.cc]: Please factor out common registrations under #if defined(INTEL_MKL) and use defined(DO_NOT_USE_ML) only for the portion that differs, that is, TF_CALL_double(...).

File: tensorflow/core/kernels/mkl_matmul_op.cc [hunk 3]
```diff
@@ -150,11 +155,25 @@ class MklMatMulOp : public OpKernel {
     // 1.0 and 0.0 respectively.
     const float alpha = 1.0f;
     const float beta = 0.0f;
+#if defined(DO_NOT_USE_ML)
+    const char* const ftrans[]   = { "N", "T", "C"};
+    int index_transa = transa? 1 : 0 ;
+    int index_transb = transb? 1 : 0 ;
+    VLOG(2) << "MKL DNN SGEMM called";
+    //Using the fortran api of mkldnn
```
Comment (tatianashp) [inline tensorflow/core/kernels/mkl_matmul_op.cc]: Is this Fortran API or C++ API expecting column major layout? Please make clear.

**Review by tatianashp — CHANGES_REQUESTED** [review]

**Review by tatianashp — APPROVED** [review]
Looks good. Thank you!

**Pull request merged**